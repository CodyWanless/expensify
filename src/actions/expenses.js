import database from '../firebase/firebase';

// ADD_EXPENSE
export const addExpense = (expense) => ({
  type: 'ADD_EXPENSE',
  expense,
});

export const startAddExpense =
  (expenseData = {}) =>
  async (dispatch) => {
    const {
      description = '',
      note = '',
      amount = 0,
      createdAt = 0,
    } = expenseData;
    const expense = {
      description,
      note,
      amount,
      createdAt,
    };

    const key = await database.insert(expense);
    return dispatch(
      addExpense({
        id: key,
        ...expense,
      }),
    );
  };

// REMOVE_EXPENSE
export const removeExpense = ({ id }) => ({
  type: 'REMOVE_EXPENSE',
  id,
});

export const startRemoveExpense =
  ({ id }) =>
  async (dispatch) => {
    await database.delete(id);
    return dispatch(removeExpense({ id }));
  };

// EDIT_EXPENSE
export const editExpense = (id, updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates,
});

export const startEditExpense = (id, updates) => async (dispatch) => {
  await database.update(id, updates);
  return dispatch(editExpense(id, updates));
};

export const setExpenses = (expenses) => ({
  type: 'SET_EXPENSES',
  expenses,
});

export const startSetExpenses = () => async (dispatch) => {
  const expenses = [];

  const snapshot = await database.getSnapshot();
  snapshot.forEach((childSnapshot) => {
    expenses.push({
      id: childSnapshot.key,
      ...childSnapshot.val(),
    });
  });

  return dispatch(setExpenses(expenses));
};
