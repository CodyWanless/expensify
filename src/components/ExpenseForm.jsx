import React from 'react';
import DatePicker from 'react-datepicker';

export default class ExpenseForm extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      description: props.expense ? props.expense.description : '',
      note: props.expense ? props.expense.note : '',
      amount: props.expense ? (props.expense.amount / 100).toString() : '',
      createdAt: props.expense ? props.expense.createdAt : new Date(),
      error: undefined,
    };
  }

  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  };

  onNoteChange = (e) => {
    const note = e.target.value;
    this.setState(() => ({ note }));
  };

  onAmountChange = (e) => {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d{1,}(.\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    }
  };

  onDateChange = (createdAt) => {
    if (createdAt) {
      this.setState(() => ({ createdAt }));
    }
  };

  saveExpense = (e) => {
    e.preventDefault();

    const { description, amount, createdAt, note } = this.state;
    const { onSubmit } = this.props;

    if (!description || !amount) {
      const error = 'Please provide description and amount.';
      this.setState(() => ({ error }));
    } else {
      this.setState(() => ({ error: '' }));
      onSubmit({
        description,
        amount: parseFloat(amount, 10) * 100,
        createdAt: createdAt.valueOf(),
        note,
      });
    }
  };

  render() {
    const { error, description, amount, note, createdAt } = this.state;
    return (
      <form className="form" onSubmit={this.saveExpense}>
        {error && <p className="form__error">{error}</p>}
        <input
          type="text"
          placeholder="Description"
          className="text-input"
          value={description}
          onChange={this.onDescriptionChange}
        />
        <input
          type="text"
          placeholder="Amount"
          className="text-input"
          value={amount}
          onChange={this.onAmountChange}
        />
        <DatePicker
          selected={createdAt}
          onChange={this.onDateChange}
          className="text-input"
          id="created-at_date-picker"
        />
        <textarea
          placeholder="Add a note for your expense (optional)"
          value={note}
          className="textarea"
          onChange={this.onNoteChange}
        />
        <div>
          <button type="submit" className="button">
            Save Expense
          </button>
        </div>
      </form>
    );
  }
}
