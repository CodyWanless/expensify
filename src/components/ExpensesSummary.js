import React from 'react';
import { connect } from 'react-redux';
import expensesTotalSelector from '../selectors/expenses-total';
import expensesSelector from '../selectors/expenses';
import numeral from 'numeral';

export const ExpensesSummary = ({expensesCount, expensesTotal }) => {
    const expensesText = expensesCount === 1 ? 'expense' : 'expenses';
    
    return (
        <div>   
            <h1>Viewing {expensesCount} {expensesText} totalling {numeral(expensesTotal / 100).format('$0,0.00')}</h1>
        </div>);
};

const mapStateToProps= (state) => {
    const expenses = expensesSelector(state.expenses, state.filters);
    
    return {
        expensesTotal: expensesTotalSelector(expenses),
        expensesCount: expenses.length
    };
};

export default connect(mapStateToProps)(ExpensesSummary);