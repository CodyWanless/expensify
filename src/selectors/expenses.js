import moment from 'moment';

export default (expenses, { text, sortBy, startDate, endDate }) =>
  expenses
    .filter((expense) => {
      const createdAtMoment = moment(expense.createdAt);
      const startDateMatch = startDate
        ? moment(startDate).isSameOrBefore(createdAtMoment, 'day')
        : true;
      const endDateMatch = endDate
        ? moment(endDate).isSameOrAfter(createdAtMoment, 'day')
        : true;
      const textMatch = expense.description
        .trim()
        .toLowerCase()
        .includes(text.trim().toLowerCase());

      return startDateMatch && endDateMatch && textMatch;
    })
    .sort((a, b) => {
      if (sortBy === 'date') {
        return a.createdAt < b.createdAt ? 1 : -1;
      }
      if (sortBy === 'amount') {
        return a.amount < b.amount ? 1 : -1;
      }
      throw new Error(`Sort by order ${sortBy} not support`);
    });
