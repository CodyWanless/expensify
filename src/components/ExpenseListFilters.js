import React from 'react';
import { connect } from 'react-redux';
import { setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate } from './../actions/filters';
import DatePicker from 'react-datepicker';

export class ExpenseListFilters extends React.Component {
    state = {
        calendarFocused: null
    };

    sortByChangedHandler = (e) => {
        const targetSortBy = e.target.value;
        if (targetSortBy === "date") {
            this.props.sortByDate();
        } else if (targetSortBy === "amount") {
            this.props.sortByAmount();
        } else {
            throw new Error("Sort by not supported");
        }
    }

    onDatesChange = ([startDate, endDate]) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    }

    onFocusChange = (calendarFocused) => {
        this.setState(() => ({
            calendarFocused
        }));
    }

    onTextChange = (e) => {
        this.props.setTextFilter(e.target.value);
    }

    render() {
        return (
            <div className="content-container">
                <div className="input-group">
                    <div className="input-group__item">
                        <input type="text"
                            className="text-input"
                            value={this.props.filters.text}
                            placeholder="Search Expenses"
                            onChange={this.onTextChange} />
                    </div>
                    <div className="input-group__item">
                        <select value={this.props.filters.sortBy}
                            onChange={this.sortByChangedHandler}
                            className="select">
                            <option value="date">Date</option>
                            <option value="amount">Amount</option>
                        </select>
                    </div>
                    <div className="input-group__item">
                        {<DatePicker
                            startDate={this.props.filters.startDate}
                            endDate={this.props.filters.endDate}
                            selectsRange={true}
                            isClearable={true}
                            onChange={this.onDatesChange}
                            onFocus={this.onFocusChange}
                            autoFocus={this.state.calendarFocused}
                            className="text-input"
                        />}
                    </div>
                </div>
            </div>
        );
    }
}

const mapDispatchToProps = (dispatch) => ({
    setTextFilter: text => dispatch(setTextFilter(text)),
    sortByDate: () => dispatch(sortByDate()),
    sortByAmount: () => dispatch(sortByAmount()),
    setStartDate: startDate => dispatch(setStartDate(startDate)),
    setEndDate: endDate => dispatch(setEndDate(endDate))
});

const mapStateToProps = (state) => ({
    filters: state.filters
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);