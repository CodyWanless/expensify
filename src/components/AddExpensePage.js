import React from 'react';
import ExpenseForm from './ExpenseForm';
import { connect } from 'react-redux';
import { startAddExpense } from './../actions/expenses';

export class AddExpensePage extends React.Component { 
    constructor(props) {
        super(props);

        this.onSubmit = this.onSubmit.bind(this);
    }

    async onSubmit(expense) {
        try {
            await this.props.startAddExpense(expense);
            this.props.history.push('/');
        } catch (e) {
            console.log(`Error adding expense: ${e}`);
        }
    }

    render() {
        return (
        <div>
            <ExpenseForm onSubmit={this.onSubmit} />
        </div>);  
    }
}

const mapDispatchToProps = (dispatch) => ({
    startAddExpense: expense => dispatch(startAddExpense(expense))
});

export default connect(undefined, mapDispatchToProps)(AddExpensePage);