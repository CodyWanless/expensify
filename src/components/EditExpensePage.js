import React from 'react';
import ExpenseForm from './ExpenseForm';
import { connect } from 'react-redux';
import { editExpense, removeExpense } from './../actions/expenses';

export class EditExpensePage extends React.Component {
    onEditExpenseSubmit = expense => {
        this.props.editExpense(this.props.expense.id, expense);
        this.props.history.push('/');
    }

    onRemoveClick = () => {
        this.props.removeExpense({ id: this.props.expense.id });
        this.props.history.push('/');
    }

    render() {
        return (
            <div>
                <ExpenseForm onSubmit={this.onEditExpenseSubmit}
                    expense={this.props.expense} />
                <button onClick={this.onRemoveClick}>Remove</button>
            </div>
        );
    }
}

const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find(e => e.id === props.match.params.id)
    };
};

const mapDispatchToProps = (dispatch) => ({
    editExpense: (id, expense) => dispatch(editExpense(id, expense)),
    removeExpense: id => dispatch(removeExpense(id))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);