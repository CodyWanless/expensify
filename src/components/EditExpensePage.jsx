import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useNavigate, useParams } from 'react-router-dom';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';
import ExpenseForm from './ExpenseForm';

export default function EditExpensePage() {
  const dispatch = useDispatch();
  const navigate = useNavigate();
  const { id } = useParams();

  const expense = useSelector((state) =>
    state.expenses.find((e) => e.id === id),
  );

  const onEditExpenseSubmit = async (expenseToAdd) => {
    await dispatch(startEditExpense(id, expenseToAdd));
    navigate('/');
  };

  const onRemoveClick = async () => {
    await dispatch(startRemoveExpense({ id }));
    navigate('/');
  };

  return (
    <div>
      <div className="page-header">
        <div className="content-container">
          <h1 className="page-header__title">Edit Expense</h1>
        </div>
      </div>
      <div className="content-container">
        <ExpenseForm onSubmit={onEditExpenseSubmit} expense={expense} />
        <button onClick={onRemoveClick} className="button button--secondary">
          Remove Expense
        </button>
      </div>
    </div>
  );
}
