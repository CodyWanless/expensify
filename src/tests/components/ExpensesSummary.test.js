import React from 'react';
import {shallow} from 'enzyme';
import { ExpensesSummary } from '../../components/ExpensesSummary';
import expenses from '../fixtures/expenses';

test ('should render ExpensesSummary correctly with 1 item', () => {
    const wrapper = shallow(<ExpensesSummary expensesCount={1} expensesTotal={4000} />);

    expect(wrapper).toMatchSnapshot();
});

test ('should render ExpensesSummary correctly with multiple items', () => {
    const wrapper = shallow(<ExpensesSummary expensesCount={3} expensesTotal={4000} />);

    expect(wrapper).toMatchSnapshot();
});