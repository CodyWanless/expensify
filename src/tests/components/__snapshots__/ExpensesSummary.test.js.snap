// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpensesSummary correctly with 1 item 1`] = `
<div>
  <h1>
    Viewing 
    1
     
    expense
     totalling 
    $40.00
  </h1>
</div>
`;

exports[`should render ExpensesSummary correctly with multiple items 1`] = `
<div>
  <h1>
    Viewing 
    3
     
    expenses
     totalling 
    $40.00
  </h1>
</div>
`;
